File: src/main/java/com/github/simiacryptus/aicoder/text/TextBlock.java
package com.github.simiacryptus.aicoder.text;

import org.jetbrains.annotations.NotNull;

import java.util.Arrays;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public interface TextBlock {

    public static final String TAB_REPLACEMENT = "  ";
    public static final String DELIMITER = "\n";

    String[] rawString();

    default String getTextBlock() {
        return Arrays.stream(rawString()).collect(Collectors.joining(DELIMITER));
    }

    @NotNull TextBlock withIndent(String indent);

    default Stream<String> stream() {
        return Arrays.stream(rawString());
    }
}

File: src/main/kotlin/com/simiacryptus/aicoder/util/TextBlock.kt
package com.simiacryptus.aicoder.util

import java.util.stream.Stream
import kotlin.streams.asStream

interface TextBlock {
  companion object {
    val TAB_REPLACEMENT: CharSequence = "  "
    const val DELIMITER: String = "\n"
  }

  fun rawString(): Array<out CharSequence>

  val textBlock: CharSequence
    get() = rawString().joinToString(DELIMITER)

  fun withIndent(indent: CharSequence): TextBlock

  fun stream(): Stream<CharSequence> = rawString().asSequence().asStream()
}